src/scripts/release.zig

```diff
@@ -530,106 +534,286 @@ fn build_tigerbeetle(shell: *Shell, info: VersionInfo, dist_dir: std.fs.Dir) !vo
+        );
+        header.current_release_client_min = (try multiversioning.Release.parse("0.15.3")).value;
+    }
 
-    const body_file = try shell.project_root.createFile(body_path, .{ .exclusive = true });
-    defer body_file.close();
+    var releases = multiversioning.ListU32{};
+    var checksums = multiversioning.ListU128{};
+    var offsets = multiversioning.ListU32{};
+    var sizes = multiversioning.ListU32{};
+    var flags_ = multiversioning.ListFlag{};
+    var git_commits = multiversioning.ListGitCommit{};
+    var release_client_mins = multiversioning.ListU32{};
+
+    // Extract the old current release - this is the release that was the current release, and not
+    // embedded in the past pack.
+    const old_current_release = header.current_release;
+    const old_current_release_output_name = try std.fmt.allocPrint(
+        shell.arena.allocator(),
+        "{s}-{}",
+        .{ tigerbeetle_name_binary, old_current_release },
+    );
 
-    try body_file.writeAll(past_binary_contents);
+    if (macos) {
+        const cpu_type = if (std.mem.eql(u8, target, "aarch64-macos"))
+            std.macho.CPU_TYPE_ARM64
+        else
+            std.macho.CPU_TYPE_X86_64;
+
+        const cpu_subtype = if (cpu_type == std.macho.CPU_TYPE_ARM64)
+            std.macho.CPU_SUBTYPE_ARM_ALL
+        else
+            std.macho.CPU_SUBTYPE_X86_64_ALL;
+
+        try macos_universal_binary_extract(
+            shell,
+            tigerbeetle_name_zip,
+            .{ .cpu_type = cpu_type, .cpu_subtype = cpu_subtype },
+            old_current_release_output_name,
+        );
+    } else {
+        try shell.exec("{llvm_objcopy} --enable-deterministic-archives --keep-undefined" ++
+            " --remove-section .tb_mvb --remove-section .tb_mvh " ++
+            "{tigerbeetle_name_zip} {old_current_release_output_name}", .{
+            .llvm_objcopy = llvm_objcopy,
+            .tigerbeetle_name_zip = tigerbeetle_name_zip,
+            .old_current_release_output_name = old_current_release_output_name,
+        });
+    }
 
-    const git_commit: [20]u8 = blk: {
-        var commit_bytes: [20]u8 = std.mem.zeroes([20]u8);
-        for (0..@divExact(multiversion_epoch_commit.len, 2)) |i| {
-            const byte = try std.fmt.parseInt(u8, multiversion_epoch_commit[i * 2 ..][0..2], 16);
-            commit_bytes[i] = byte;
+    // It's important to verify the previous current_release checksum - it can't be verified at
+    // runtime by multiversioning.zig, since it relies on objcopy to extract.
+    assert(header.current_checksum == try checksum_file(
+        shell,
+        old_current_release_output_name,
+        multiversion_binary_size_max,
+    ));
+
+    const old_current_release_size: u32 = @intCast((try (try shell.cwd.openFile(
+        old_current_release_output_name,
+        .{ .mode = .read_only },
+    )).stat()).size);
+
+    // FIXME: This can vary by platform; so we could create a binary for Linux that has 10 past
```

Comment:
An unsolved problem here. We could:
* make it assert the counts are identical, and fix it down the line,
* limit to the lessor of all platforms
* ???

The debug builds are very large - 20MB a pop on Linux. With the doubled max of 128MB, that's 6 releases. For production, the 64MB max could handle 23 releases at ~2.7MB each.

Reply:
Hm, this entire logic seems backwards (modulo my current _very_ foggy understanding of the logic, might be talking extreme amount of nonsense here)

It looks like what we do here is that we have a certain size budget, and try to fit as many versions as possible. 

I'd expect this to go the other way around: just _declare_ that "I want exactly X versions", pack the X, and then asser that it fits the size budget?

Reply:
> I'd expect this to go the other way around: just declare that "I want exactly X versions", pack the X, and then asser that it fits the size budget?

We have two limits: a logical number on the past releases `past_releases_max` (which is `constants.vsr_releases_max - 1`, ie 64) and the physical binary size we're willing to put out (because we have to allocate for it, it's not free to increase. These numbers were mostly plucked from "how big of a binary do we feel comfortable shipping". 64MB on Windows / Linux, x2 for Mac, further x2 for debug).

This leads to an implicit number of releases, but that was by design because the size of a releases is not fixed (and we might introduce compression). In general, we want to pack as many as possible in.

That said, this is the release logic, so this is all easy to change. We could define targets, assuming a certain size per build, and assert that those are under the maximum size. If we ever hit that assert in a build, we deal with it then by lowering the target, reducing binary size, etc.

Reply:
>This leads to an implicit number of releases, but that was by design because the size of a releases is not fixed (and we might introduce compression). In general, we want to pack as many as possible in

Hm, this does look surprising. If I am a user, I probably have some sort of automation running to pick new releases automatically, and that automation probably would want to depend on specific number of releases available. 

I am worrying about the scenarios where we always packed 8 releases, but then one day we blew up the code somewhere, packed 7 by accident, and somoene eles rollout blew up.

Reply:
> Hm, this does look surprising. If I am a user, I probably have some sort of automation running to pick new releases automatically, and that automation probably would want to depend on specific number of releases available.

Hmm - I'm not sure that's going to be common :thinking:. We could perhaps drop releases for other reasons, and I think we would expect users to have fixed versions installed and not have new releases automatically installed.

The release notes would described the minimum supported versions for both the replica and client_release_min.

Reply:
:thinking: yeah, I am not sure what we want to do here. I think I am still more worried about the count being different between two subsequent linux releases, then between a linux and a windows release. 

My instinct here is just say that we always pack just `4` releases, and then adjust as we go. 

In particular, it seems to me that we won't actually want to pack consequtive releases, if we have `visit` infra? I'd expect us to do some kind of binary climbing, where you have

* the latest release
* the release before that
* the previous release that's divisible by 4
* the previous release that's divisible by 8
* the previous release that's divisible by 16

such that you can jump over large slices of time...

Reply:
> 🤔 yeah, I am not sure what we want to do here. I think I am still more worried about the count being different between two subsequent linux releases, then between a linux and a windows release.

Discussed this today, one idea was to flip it around:
* Binaries get a fixed size cap, eg 4MB for release, 20MB for debug
* The number we pack in gets a fixed multiple, eg 10 binaries
* That determines our memory limit: 40MB release, 200MB debug

Another idea was to remove some of the TOCTOU proofing, and get rid of the double buffering and exec'ing the latest binary from memfd rather than disk, so we'd only need to statically allocate the memory for 1 release rather than the whole binary size.

> My instinct here is just say that we always pack just 4 releases, and then adjust as we go.

I think for this version, let's do exactly that.

> In particular, it seems to me that we won't actually want to pack consequtive releases, if we have visit infra? I'd expect us to do some kind of binary climbing, where you have

That's clever, but you'd have to couple that with an intermediate to first get you to that previous release divisible by x number (the binary on disk must contain the running version, even if it's not marked as visit).

Reply:
> Hm, one complication here would be that we won't know these versions until after the release is done? We write notes first ,and then make a release, and only during release we'd figure out how much of the past releases we actually were able to fit?

Not quite, I was thinking [here](https://github.com/tigerbeetle/tigerbeetle/releases/tag/0.15.4) which is generated after the final binaries have been generated.